xdem/coreg/affine.py

```diff
@@ -740,282 +782,1289 @@ def vertical_shift(
+
+    :return: Affine transform matrix, Centroid, Subsample size.
+    """
+
+    # Derive normals if method is point-to-plane, otherwise not
+    if method == "point-to-plane":
+        # We use the DEM to derive the normals
+        if isinstance(ref_elev, np.ndarray):
+            dem = ref_elev
+        else:
+            dem = tba_elev
+        nx, ny, nz = _icp_norms(dem, transform)
+        aux_vars = {"nx": nx, "ny": ny, "nz": nz}
+    else:
+        aux_vars = None
+
+    # Pre-process point-raster inputs to the same subsampled points
+    sub_ref, sub_tba, sub_aux_vars, sub_coords = _preprocess_pts_rst_subsample(
+        params_random=params_random,
+        ref_elev=ref_elev,
+        tba_elev=tba_elev,
+        inlier_mask=inlier_mask,
+        transform=transform,
+        crs=crs,
+        area_or_point=area_or_point,
+        z_name=z_name,
+        aux_vars=aux_vars,
+        return_coords=True,
+    )
+
+    # Convert point clouds to Nx3 arrays for efficient calculations below
+    ref_epc = np.vstack((sub_coords[0], sub_coords[1], sub_ref))
+    tba_epc = np.vstack((sub_coords[0], sub_coords[1], sub_tba))
+    if sub_aux_vars is not None:
+        norms = np.vstack((sub_aux_vars["nx"], sub_aux_vars["ny"], sub_aux_vars["nz"]))
+    else:
+        norms = None
+
+    # Remove centroid and standardize to facilitate numerical convergence
+    ref_epc, tba_epc, centroid, std_fac = _standardize_epc(ref_epc, tba_epc, scale_std=standardize)
+    tolerance /= std_fac
+
+    # Define search tree outside of loop for performance
+    ref_epc_nearest_tree = scipy.spatial.KDTree(ref_epc.T)
+
+    # Iterate through method until tolerance or max number of iterations is reached
+    init_matrix = np.eye(4)  # Initial matrix is the identity transform
+    constant_inputs = (
+        ref_epc,
+        tba_epc,
+        norms,
+        ref_epc_nearest_tree,
+        params_fit_or_bin,
+        method,
+        picky,
+        only_translation,
+    )
+    final_matrix = _iterate_method(
+        method=_icp_iteration_step,
+        iterating_input=init_matrix,
+        constant_inputs=constant_inputs,
+        tolerance=tolerance,
+        max_iterations=max_iterations,
+    )
+    # De-standardize
+    final_matrix[:3, 3] *= std_fac
+
+    # Get subsample size
+    subsample_final = len(sub_ref)
+
+    return final_matrix, centroid, subsample_final
+
+
+#########################
+# 5/ Coherent Point Drift
+#########################
+
+
+def _cpd_fit(
+    ref_epc: NDArrayf,
+    tba_epc: NDArrayf,
+    trans_tba_epc: NDArrayf,
+    weight_cpd: float,
+    sigma2: float,
+    sigma2_min: float,
+    scale: bool = False,
+    only_translation: bool = False,
+) -> tuple[NDArrayf, float, float]:
+    """
+    Fit step of Coherent Point Drift by expectation-minimization, with variance updating.
+
+    See Fig. 2 of Myronenko and Song (2010), https://arxiv.org/pdf/0905.2635.pdf for equations below.
```

Comment:
Should use the DOI: [10.1109/TPAMI.2010.46](https://doi.org/10.1109/TPAMI.2010.46)